File: packages/docs/make-reference.py
"""Copy generated API docs and build sidebar for Starlight."""

import json
import re
import shutil
from collections import defaultdict
from pathlib import Path

DOCS_DIR = Path(__file__).resolve().parent
GENERATED_DIR = DOCS_DIR / "../cmaf-lite/__generated_docs__"
MD_SOURCE_DIR = GENERATED_DIR / "markdown"
API_JSON_PATH = GENERATED_DIR / "cmaf-lite.api.json"
REF_DIR = DOCS_DIR / "src/content/docs/reference"

# Copy markdown files
if REF_DIR.exists():
    shutil.rmtree(REF_DIR)
shutil.copytree(MD_SOURCE_DIR, REF_DIR)

# Build sidebar from api.json
api_model = json.loads(API_JSON_PATH.read_text())
pkg = api_model["members"][0]
pkg_name = api_model["name"]

kind_groups = defaultdict(list)
for member in pkg.get("members", []):
    name = member.get("name", "")
    link = f"/reference/{pkg_name}.{name}/".lower()
    md_name = f"{pkg_name}.{name}".lower() + ".md"
    if md_name.endswith("event.md"):
        kind_groups["Event"].append({"label": name, "link": link})
    else:
        kind_groups[member["kind"]].append({"label": name, "link": link})

sidebar = sorted(
    [
        {"label": kind, "items": sorted(items, key=lambda x: x["label"])}
        for kind, items in kind_groups.items()
    ],
    key=lambda x: x["label"],
)
for group in sidebar[1:]:
    group["collapsed"] = True

# Add frontmatter and rewrite links
KIND_RE = re.compile(r" (class|enum|interface|variable|type alias|type|package)$", re.I)
LINK_RE = re.compile(r"\./(\w[\w.-]+)\.md")

for md in REF_DIR.glob("*.md"):
    content = md.read_text()
    title_match = re.search(r"^## (.+)$", content, re.M)
    title = KIND_RE.sub("", title_match.group(1)) if title_match else md.name
    slug = f"reference/{md.stem}"
    rewritten = LINK_RE.sub(r"/reference/\1", content)
    md.write_text(f"---\ntitle: {title}\nslug: {slug}\n---\n\n{rewritten}")

# Write sidebar
sidebar_path = DOCS_DIR / "sidebar-reference.json"
sidebar_json = [{"label": "Reference", "collapsed": True, "items": sidebar}]
sidebar_path.write_text(json.dumps(sidebar_json, indent=2))

File: packages/docs/scripts/make-reference.ts
import { mkdirSync, readdirSync, readFileSync, writeFileSync } from "node:fs";
import { resolve } from "node:path";

const docsDir = resolve(import.meta.dirname, "..");
const generatedDir = resolve(docsDir, "../cmaf-lite/__generated_docs__");
const mdSourceDir = resolve(generatedDir, "markdown");
const apiJsonPath = resolve(generatedDir, "cmaf-lite.api.json");
const refDir = resolve(docsDir, "src/content/docs/reference");

interface ApiMember {
  kind: string;
  name?: string;
  members?: ApiMember[];
}

/**
 * Copy generated markdown files into the reference
 * directory.
 */
function copyMarkdown() {
  mkdirSync(refDir, { recursive: true });
  for (const file of readdirSync(mdSourceDir)) {
    writeFileSync(
      resolve(refDir, file),
      readFileSync(resolve(mdSourceDir, file)),
    );
  }
}

/**
 * Build sidebar groups from the api.json file.
 */
function buildSidebar() {
  const apiModel = JSON.parse(readFileSync(apiJsonPath, "utf-8"));
  const pkg = apiModel.members[0] as ApiMember;
  const pkgName = apiModel.name as string;

  const kindGroups = new Map<string, { label: string; link: string }[]>();
  for (const member of pkg.members ?? []) {
    const name = member.name ?? "";
    const link = `/reference/${pkgName}.${name}/`.toLowerCase();
    const items = kindGroups.get(member.kind) ?? [];
    items.push({ label: name, link });
    kindGroups.set(member.kind, items);
  }

  return Array.from(kindGroups, ([kind, items]) => ({
    label: kind,
    collapsed: true,
    items: items.sort((a, b) => a.label.localeCompare(b.label)),
  }));
}

/**
 * Add frontmatter and rewrite links for all markdown
 * files in refDir.
 */
function addFrontmatter() {
  for (const file of readdirSync(refDir).filter((f) => f.endsWith(".md"))) {
    const filePath = resolve(refDir, file);
    const content = readFileSync(filePath, "utf-8");

    const rawTitle = content.match(/^## (.+)$/m)?.[1] ?? file;
    const title = rawTitle.replace(
      / (class|enum|interface|variable|type alias|type|package)$/i,
      "",
    );
    const slug = `reference/${file.replace(".md", "")}`;
    const rewritten = content.replaceAll(/\.\/([\w.-]+)\.md/g, "/reference/$1");

    writeFileSync(
      filePath,
      `---\ntitle: ${title}\nslug: ${slug}\n---\n\n${rewritten}`,
    );
  }
}

// Run pipeline
copyMarkdown();

const sidebar = [
  { label: "cmaf-lite", items: buildSidebar() },
];

addFrontmatter();

writeFileSync(
  resolve(docsDir, "sidebar-reference.json"),
  JSON.stringify([{ label: "Reference", items: sidebar }], null, 2),
);
